### PR TITLE
CIP 4: Proposal for hotfixes via governance

### PR DESCRIPTION
**Pull request by timmoreton**
Proposal aimed at striking a balance between the need for community governance and the need for a rapid and discreet response to critical security vulnerabilities.

### CODE REVIEW

**Comment by yorhodes** [comment]
WIP implementation being worked on in https://github.com/celo-org/celo-monorepo/pull/613

File: CIPs/0004.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+## Proposed Solution
+
+Zero-day vulnerability reports are directed off-chain to the Celo Foundation, which coordinates with committers to verify the issue and build a hotfix. They show the hotfix to a trusted third-party auditor, likely a reputable security firm, to verify its legitimacy and that the patch does indeed fix the issue. With the auditor’s attestation, they reach out to the validators off-chain using metadata managed as proposed in [CIP 3](0003.md) with the auditor’s attestation to convince them to apply the hotfix.
+
+### Hotfixes for Celo Core Contracts
+
+A hotfix for vulnerabilities in the Celo smart contracts takes the form of a governance proposal, but it does not follow the standard proposal procedure. 
+
+Once a validator is convinced that the hotfix is legitimate, the validator calls a function in Governance to whitelist the hotfix’s hash. The validity of the hash is also corroborated by the auditor, as the validator is not shown the plaintext hotfix. When `2f+1` of the validators have whitelisted the hash, the hotfix can be executed through Governance immediately.
```
[the remark this answers is not in the document]
Comment (timmoreton) [reply CIPs/0004.md]: addressed by @yorhodes' edits

File: CIPs/0004.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+Each validator as well as the `approvers` address can call `whitelist(hash)` in Governance to add their address for a given hash. 
+
+Calling `hotfix(proposal, hash)` with a proposal checks the hash against the proposal, then checks the whitelist for the hash. If the `approvers` address has whitelisted the hash, then it checks how many currently elected validators have whitelisted it. If the count exceeds the Byzantine Quorum `2f+1` given the current `maxElectableValidators`, the proposal is executed.
```
[the remark this answers is not in the document]
Comment (timmoreton) [reply CIPs/0004.md]: you're absolutely right. Updated.

File: CIPs/0004.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+To minimize single-epoch coordination amongst validators, whitelists are persistent across epochs. This design decision requires a whitelist tally to be computed for a specific epoch (and correspondingly a given set of validators). 
+
+Calling `prepareHotfix(hash)` computes this tally from the current validator set, and if the count exceeds the Byzantine Quorum `2f+1` given the current `numberValidatorsInCurrentSet`, marks the hash prepared for the current epoch. 
+
+Calling `executeHotfix(proposal)` computes the hash of a proposal. If the hash...
+- has been whitelisted by the `approver`
+- is marked prepared for the current epoch 
+- is not marked executed
+
+then the proposal is executed and the hash is marked executed accordingly.
```
Comment (timmoreton) [inline CIPs/0004.md]: formatting nit: in the md this gets tacked on to the final bullet?

File: CIPs/0004.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+* Validators are unreachable in a timely fashion - mitigated via [CIP 3](0003.md)
+* Validators cannot be convinced of the veracity of a fix - mitigated by Celo Foundation enlisting auditors and publishing their public keys in a well-known location, and auditors signing the report linking a vulnerability to a proposal hash.
+* Validators collude to modify the network -- in any case, BFT assumes `2f+1` honest validators.
+* Committers and auditors collude -- Celo Foundation should vet and appoint auditors. Foundation should be involved in teraction between auditors and committers. `approvers` multisig must be set such that individual committers cannot whitelist proposals. 
+
+## Implementation
+
+* To follow.
+
+g8t
```
Comment (timmoreton) [inline CIPs/0004.md]: cut!

**Pull request merged**